### PR TITLE
Add "hard reset" operation

### LINKED ISSUE #7261: Add API to move branch head to a commit (`git reset --hard`)
Add an API to set a branch head to an arbitrary commit.

This is a potentially dangerous feature, especially given that lakeFS has no
reflog (#2321).  So no UI (CLI or GUI) for the time being.

### PR DESCRIPTION
**Pull request by arielshaqed**
Fixes #7261 .

### CODE REVIEW

File: pkg/graveler/graveler.go [hunk 1]
```diff
@@ -2226,6 +2229,68 @@ func (g *Graveler) dropTokens(ctx context.Context, tokens ...StagingToken) {
...
+		// through atomic operations.
+		//
+		// So we need to dereference here :-(
+		commitRecord, err := g.dereferenceCommit(ctx, repository, ref)
+		if err != nil {
+			return nil, err
+		}
+		if branch.StagingToken != "" || len(branch.SealedTokens) > 0 {
+			return nil, ErrDirtyBranch
+		}
```
Comment (nopcoder) [inline pkg/graveler/graveler.go:2287]: - Chekcing staging token before dereference will do better
- We have isSealedEmpty

File: pkg/graveler/graveler.go [hunk 2]
```diff
@@ -2226,6 +2229,68 @@ func (g *Graveler) dropTokens(ctx context.Context, tokens ...StagingToken) {
...
+	for _, opt := range opts {
+		opt(options)
+	}
+
+	if repository.ReadOnly && !options.Force {
+		return ErrReadOnlyRepository
+	}
+
+	// TODO(ariels): up to here.  Verify staging is empty!
+	err = g.RefManager.BranchUpdate(ctx, repository, branchID, func(branch *Branch) (*Branch, error) {
```
Comment (nopcoder) [inline pkg/graveler/graveler.go:2258]: We can use `retryBranchUpdate`

File: esti/rollback_test.go [hunk 3]
```diff
@@ -4,55 +4,79 @@ import (
...
+
+			// read file
+			getObjResp, err := client.GetObjectWithResponse(ctx, repo, mainBranch, &apigen.GetObjectParams{Path: objPath})
+			require.NoError(t, err, "failed to get object")
+			require.NoErrorf(t, verifyResponse(getObjResp.HTTPResponse, getObjResp.Body),
+				"failed to get object repo %s branch %s path %s", repo, mainBranch, objPath)
+
+			// assert file content
+			body := string(getObjResp.Body)
+			require.Equal(t, objContent, body, fmt.Sprintf("path: %s, expected: %s, actual:%s", objPath, objContent, body))
```
Comment (nopcoder) [inline esti/rollback_test.go:67]: Equalf can be use to format the message

**Review by nopcoder — CHANGES_REQUESTED** [review]
Small comments on graveler.go

File: pkg/graveler/graveler.go (hunk 1 above)
Comment (arielshaqed) [reply pkg/graveler/graveler.go]: Thanks, done!

**Review by arielshaqed — COMMENTED** [review]
Thanks!

PTAL: I changed the API _completely_ to use a different endpoint.  Otherwise the error checking was incredibly complicated and fragile: for instance, for a hard reset we needed to _forbid_ `type`, which is a _required_ `enum` for the non-hard reset.

**Review by nopcoder — APPROVED** [review]
LGTM!

**Comment by idanovo** [comment]
LGTM
Missing two things:
1. Test we can't hard reset a branch of a read-only repository
2. lakectl

**Review by N-o-Z — COMMENTED** [review]
@arielshaqed 
Can we please tag this API as internal?
My concern is that after releasing it as a public API, we will not be able to remove it (if we eventually decide to) due to our API guarantees.
We can always decide to move it from internal to external if we decide we actually want to support this feature

**Comment by idanovo** [comment]
> @arielshaqed Can we please tag this API as internal? My concern is that after releasing it as a public API, we will not be able to remove it (if we eventually decide to) due to our API guarantees. We can always decide to move it from internal to external if we decide we actually want to support this feature

@N-o-Z @arielshaqed 
OK, if we decide to make it internal we probably don't need a lakectl command.
I'm not sure it should be an internal API, maybe we need to introduce a new "experimental" flag?

**Comment by N-o-Z** [comment]
> > @arielshaqed Can we please tag this API as internal? My concern is that after releasing it as a public API, we will not be able to remove it (if we eventually decide to) due to our API guarantees. We can always decide to move it from internal to external if we decide we actually want to support this feature
> 
> @N-o-Z @arielshaqed OK, if we decide to make it internal we probably don't need a lakectl command. I'm not sure it should be an internal API, maybe we need to introduce a new "experimental" flag?

As far as I understood from @arielshaqed no UX means no lakectl as well.
What is your concern about making it internal? It is just a tag which tells the user that they should not use this API and if they do they do it at their own risk.
Experimental in my opinion is by definition something transient, and we already know this is required for replication

**Comment by idanovo** [comment]
> > > @arielshaqed Can we please tag this API as internal? My concern is that after releasing it as a public API, we will not be able to remove it (if we eventually decide to) due to our API guarantees. We can always decide to move it from internal to external if we decide we actually want to support this feature
> > 
> > 
> > @N-o-Z @arielshaqed OK, if we decide to make it internal we probably don't need a lakectl command. I'm not sure it should be an internal API, maybe we need to introduce a new "experimental" flag?
> 
> As far as I understood from @arielshaqed no UX means no lakectl as well. What is your concern about making it internal? It is just a tag which tells the user that they should not use this API and if they do they do it at their own risk. Experimental in my opinion is by definition something transient, and we already know this is required for replication

@N-o-Z I think it makes sense users will want to use this API so can't see why it should be an internal API.
If we think we might want to remove the support in the future, experimental is more suitable  IMO.

**Comment by arielshaqed** [comment]
> @arielshaqed Can we please tag this API as internal? My concern is that after releasing it as a public API, we will not be able to remove it (if we eventually decide to) due to our API guarantees. We can always decide to move it from internal to external if we decide we actually want to support this feature

You're right!  So I tagged it "experimental" since there are many use-cases for it, we've even seen requests.

**Comment by arielshaqed** [comment]
@idanovo wrote:
>  I'm not sure it should be an internal API, maybe we need to introduce a new "experimental" flag?

I looked at the Swagger file and I cannot introduce a new "experimental" tag _because we already have one_ 😇 .  So I'm using that one!

**Comment by N-o-Z** [comment]
> @idanovo wrote:
> 
> > I'm not sure it should be an internal API, maybe we need to introduce a new "experimental" flag?
> 
> I looked at the Swagger file and I cannot introduce a new "experimental" tag _because we already have one_ 😇 . So I'm using that one!

Some questions about the `experimental` tag:
1. What makes this API experimental?
2. What are the restrictions for using experimental APIs (have we communicated it to our users)?
3. What are our declared guarantees for them?

**Comment by arielshaqed** [comment]
@idanovo wrote:

> LGTM Missing two things:
> 
>     1. Test we can't hard reset a branch of a read-only repository

I couldn't find where we do this for _any_ of the Reset calls.  Opened #7266.

>     2. lakectl

Now that it's "experimental" I think we can definitely avoid it.  In any case, as a user, I would be _very_ afraid to use a hard-reset command when there it no "undo".  So I am opposed to adding a direct command for users to do this until we have a reflog (#2321, #6373).

**Comment by arielshaqed** [comment]
> > @idanovo wrote:
> > > I'm not sure it should be an internal API, maybe we need to introduce a new "experimental" flag?
> > 
> > 
> > I looked at the Swagger file and I cannot introduce a new "experimental" tag _because we already have one_ 😇 . So I'm using that one!
> 
> Some questions about the `experimental` tag:
> 
>     1. What makes this API experimental?

We don't know whether it will stay or be replaced with another API or get some new parameters.  It is unstable.  Making it "experimental" is the way to do [this](https://github.com/treeverse/lakeFS/pull/7263#pullrequestreview-1815073520), without stating that it is "internal".

>     2. What are the restrictions for using experimental APIs (have we communicated it to our users)?
> 
>     3. What are our declared guarantees for them?

[This section](https://docs.lakefs.io/understand/towards-1.0-sdk.html#interface-guarantees) defines (no...) interface guarantees for "internal" and "experimental" APIs.  We do not define a clear difference between the two.  However I believe that developers will regard "internal" as "lakeFS devs may use this to do something, and you really cannot expect any stability here" and "experimental" as "this does something, it may become a stable API, and you are welcome to try it out until it does".

**Pull request merged**